src/workerd/io/worker.c++

```diff
@@ -2327,8 +2311,236 @@ public:
     // delay signaling the outgoing loop until this call?
   }
 
+  void pauseIncomingMessages();
+  // Stops the delivery of CDP messages on the I/O worker thread. Called when the isolate hits a
+  // breakpoint or debugger statement.
+
+  void resumeIncomingMessages();
+  // Resumes delivery of CDP messages on the I/O worker thread. Called when execution resumes after
+  // a breakpoint or debugger statement.
+
+  bool dispatchOneMessageDuringPause();
+  // Dispatches one message whilst automatic CDP messages on the I/O worker thread is paused, called
+  // on the thread executing the isolate whilst execution is suspended due to a breakpoint or
+  // debugger statement.
+
 private:
-  kj::WebSocket& webSocket;
+  class WebSocketIoHandler final {
+    // Class that manages the I/O for devtools connections. I/O is performed on the
+    // thread associated with the InspectorService (the thread that calls attachInspector).
+    // Most of the public API is intended for code running on the isolate thread, such as
+    // the InspectorChannelImpl and the InspectorClient.
+  public:
+    WebSocketIoHandler(kj::WebSocket& webSocket)
+        : webSocket(webSocket), receivedClose(false) {
+      // Assume we are being instantiated on the InspectorService thread, the thread that will do
+      // I/O for CDP messages. Messages are delivered to the InspectorChannelImpl on the Isolate thread.
+      outgoingQueueNotifier = kj::atomicRefcounted<XThreadNotifier>(kj::getCurrentThreadExecutor());
+    }
+
+    ~WebSocketIoHandler() noexcept(false) {
+      outgoingQueueNotifier->clear();
+    }
+
+    void connect(InspectorChannelImpl& inspectorChannel) {
+      // Sets the channel that messages are delivered to.
+      channel = inspectorChannel;
+    }
+
+    void disconnect() {
+      channel = {};
+      shutdown();
+    }
+
+    bool isClosed() const {
+      return receivedClose.load(std::memory_order_acquire);
+    }
+
+    void pauseIncomingMessages() {
+      auto lockedIncomingQueue = incomingQueue.lockExclusive();
+      lockedIncomingQueue->paused = true;
+    }
+
+    void resumeIncomingMessages() {
+      auto lockedIncomingQueue = incomingQueue.lockExclusive();
+      lockedIncomingQueue->paused = false;
+      deliverProtocolMessages(channel, lockedIncomingQueue);
+    }
+
+    kj::Maybe<kj::String> waitForMessage() {
+      // Blocked the current thread until a message arrives. This is intended
+      // for use in the InspectorClient when breakpoints are hit. The InspectorClient
+      // has to remain in runMessageLoopOnPause() but still receive CDP messages
+      // (e.g. resume).
+      return incomingQueue.when(
+        [this](const MessageQueue& incomingQueue) {
+          return (incomingQueue.head < incomingQueue.messages.size() || isClosed());
```

Comment:
The comments I tried to post twice before and github lost it both times, it seems:

It feels weird that this conditional wait is considering `isClosed()`, which isn't part of the mutex state. The condition won't automatically be checked when `isClosed()` changes state, it's only checked when the mutex is unlocked implying that something could have changed. I see that in practice you always lock and unlock this mutex shortly after changing the isClosed() flag, so it works out in practice. But, perhaps it would be less brittle to make the closed state be part of the `incomingQueue` mutex?

Note that if you need to access it atomically somewhere where locking the mutex would be too much overhead, you can use `getWithoutLock()` to skip the lock (and this should be safe enough when atomically reading a boolean).

Reply:
Thanks, added a status field to the queue definition (active, paused, closed).